### PR TITLE
Implements tests for MarkdownParser and fixes some bugs related to its

### PR DESCRIPTION
**Pull request by BarkingBad**

### CODE REVIEW

File: core/src/main/kotlin/parsers/MarkdownParser.kt [hunk 1]
```diff
@@ -77,34 +79,63 @@ class MarkdownParser (
...
-            } else {
-                    resolveKDocLink(
-                        resolutionFacade.resolveSession.bindingContext,
-                        resolutionFacade,
-                        declarationDescriptor,
-                        null,
-                        link.split('.')
-                    ).also { if (it.size > 1) DokkaConsoleLogger.warn("Markdown link resolved more than one element: $it") }.firstOrNull()//.single()
-                        ?.let { DRI.from(it) }
+        private fun resolveDRI(link: String): DRI? = if (link.startsWith("http") || link.startsWith("www")) {
```
Comment (kamilok1965) [inline core/src/main/kotlin/parsers/MarkdownParser.kt]: This should be done in another method I think, as the list of URI prefixes can be longer (`file`, `ftp`, `mailto` and so on). Moreover I think we could recognise it somehow (maybe by a colon after the first word or by creating a `java.net.URL` and catching `MalformedURLException`)

File: core/src/main/kotlin/parsers/MarkdownParser.kt [hunk 2]
```diff
@@ -171,20 +207,84 @@ class MarkdownParser (
                 MarkdownElementTypes.IMAGE                  -> imagesHandler(node)
                 MarkdownTokenTypes.CODE_FENCE_CONTENT,
                 MarkdownTokenTypes.CODE_LINE,
-                MarkdownTokenTypes.TEXT                     -> DocNodesFromIElementFactory.getInstance(MarkdownTokenTypes.TEXT, body = text.substring(node.startOffset, node.endOffset))
+                MarkdownTokenTypes.TEXT                     -> DocNodesFromIElementFactory.getInstance(
+                    MarkdownTokenTypes.TEXT,
+                    body = text
+                        .substring(node.startOffset, node.endOffset).transform()
```
Comment (kamilok1965) [inline core/src/main/kotlin/parsers/MarkdownParser.kt]: Could you please add a comment in code explaining why this is doing what this is doing?

File: core/src/main/kotlin/parsers/MarkdownParser.kt [hunk 3]
```diff
@@ -171,20 +207,84 @@ class MarkdownParser (
...
+                    }
+                }
+                index++
+            }
+            return children
+        }
+
+        private fun String.transform() = this
+            .replace(Regex("\n\n+"), "")
+            .replace(Regex("\n>+ "), "\n")
```
Comment (kamilok1965) [inline core/src/main/kotlin/parsers/MarkdownParser.kt]: May Windows line endings be present there? If so they should also be replaced if I understand this correctly

File: testApi/src/main/kotlin/testApi/testRunner/TestRunner.kt [hunk 4]
```diff
@@ -197,6 +200,48 @@ abstract class AbstractCoreTest {
     }
 }
 
+abstract class AbstractKDocTest : AbstractCoreTest() {
```
Comment (kamilok1965) [inline testApi/src/main/kotlin/testApi/testRunner/TestRunner.kt]: IMHO this should not be in the testAPI as those methods are meant for a specific type of KDoc tests. That's why this should be moved to a separate file in core/test/markdown

File: core/src/main/kotlin/parsers/MarkdownParser.kt (hunk 1 above)
Comment (BarkingBad) [reply core/src/main/kotlin/parsers/MarkdownParser.kt]: Second option seems nice

Comment (BarkingBad) [reply core/src/main/kotlin/parsers/MarkdownParser.kt]: Done

File: core/src/main/kotlin/parsers/MarkdownParser.kt (hunk 3 above)
Comment (BarkingBad) [reply core/src/main/kotlin/parsers/MarkdownParser.kt]: Tested on CRLF, did fine. We can add in the future source file test with CRLF encoded file to automate and ensure this case

File: testApi/src/main/kotlin/testApi/testRunner/TestRunner.kt (hunk 4 above)
Comment (BarkingBad) [reply testApi/src/main/kotlin/testApi/testRunner/TestRunner.kt]: Moved

**Comment by kamilok1965** [comment]
One more thing, parser should probably throw something more specific than just `Error`s

**Comment by BarkingBad** [comment]
> One more thing, parser should probably throw something more specific than just `Error`s

Perfectly, it should be handled by error logger, but I wasn't sure can I make parsers dependent from logger/context

**Pull request merged**